app/src/main/java/org/mozilla/focus/activity/MainMediator.java

```diff
@@ -98,90 +92,10 @@ boolean shouldFinish() {
         }
 
         FragmentManager.BackStackEntry lastEntry = manager.getBackStackEntryAt(entryCount - 1);
-        return TYPE_ROOT.equals(lastEntry.getName());
-    }
-
-    private void clearBackStack(FragmentManager fm) {
-        fm.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);
-    }
-
-    private void clearAllFragmentImmediate() {
-        final FragmentManager fragmentMgr = this.activity.getSupportFragmentManager();
-        if (fragmentMgr.isStateSaved()) {
-            return;
-        }
-
-        final Fragment urlInputFrg = fragmentMgr.findFragmentByTag(UrlInputFragment.FRAGMENT_TAG);
-        final Fragment homeFrg = fragmentMgr.findFragmentByTag(HomeFragment.FRAGMENT_TAG);
-
-        // If UrlInputFragment exists, remove it and clear its transaction from back stack
-        FragmentTransaction clear = fragmentMgr.beginTransaction();
-        clear = (urlInputFrg == null) ? clear : clear.remove(urlInputFrg);
-        clear = (homeFrg == null) ? clear : clear.remove(homeFrg);
-        clear.commit();
-
-        clearBackStack(fragmentMgr);
-    }
-
-    void clearAllFragment(boolean animate) {
-        final FragmentManager manager = this.activity.getSupportFragmentManager();
-        final Fragment homeFragment = manager.findFragmentByTag(HomeFragment.FRAGMENT_TAG);
-        if (!animate || homeFragment == null) {
-            clearAllFragmentImmediate();
-            return;
-        }
-
-        fadeOutFragment(homeFragment, this::clearAllFragmentImmediate);
-    }
-
-    private void fadeOutFragment(Fragment fragment, @Nullable final Runnable onAnimationEndCallback) {
-        View view = fragment.getView();
-        if (view == null) {
-            if (onAnimationEndCallback != null) {
-                onAnimationEndCallback.run();
-            }
-            return;
-        }
-
-        Animation currentAnim = view.getAnimation();
-        if (currentAnim != null && !currentAnim.hasEnded()) {
-            return;
-        }
-
-        Animation fadeOut = AnimationUtils.loadAnimation(view.getContext(),
-                R.anim.tab_transition_fade_out);
-        fadeOut.setAnimationListener(new Animation.AnimationListener() {
-            @Override
-            public void onAnimationStart(Animation animation) {
-            }
-
-            @Override
-            public void onAnimationEnd(Animation animation) {
-                if (onAnimationEndCallback != null) {
-                    onAnimationEndCallback.run();
-                }
-            }
-
-            @Override
-            public void onAnimationRepeat(Animation animation) {
-            }
-        });
-        view.startAnimation(fadeOut);
-    }
-
-    void onFragmentStarted(@NonNull String tag) {
-        if (UrlInputFragment.FRAGMENT_TAG.equals(tag)) {
```

Comment:
I'm a bit lost here. Why don't we just do toggleFakeUrlInput() in UrlInputFragment's onStart?

Reply:
I don't think UrlInputFragment should know when or how a specific UI element on home fragment would behave.